### ISSUE TITLE
Scandinavian keyboard layout can't handle input tilde (~)

### ISSUE THREAD

**Issue by ixrock** [xtermjs/xterm.js#2151]
To type `~` with Finnish/Swedish keyboard I have to press hotkey `alt+^`:
![image](https://user-images.githubusercontent.com/6377066/58760999-9f1a8b80-8547-11e9-8c33-f75d9682984c.png)

After that it shows something similar to `~` but right after when I press any other character (except Enter) it will erase "pre-tilde" symbol. Press `Enter` "converts" to normal ~ and nothing more (without real Enter).

I checked how this works with vanilla JS and it seems something wrong at browser-level:
![image](https://user-images.githubusercontent.com/6377066/58705813-f0e0db80-83b8-11e9-8104-50cb617dd00c.png)

Is it fixable? Wdyt?

## Related bugs:
https://github.com/xtermjs/xterm.js/issues/174

## Details
- Browser and browser version: `Chrome (latest)`
- OS version: `MacOS X (latest)`
- xterm.js version: `3.14.0`

### Steps to reproduce
1. Add  support of Finnish keyboard in your OS
2. Try to use ~ from keyboard

**Comment by Tyriar** [comment]
VS Code issue: https://github.com/microsoft/vscode/issues/68262

**Comment by ixrock** [comment]
@Tyriar how `xterm` related to `vscode`?

**Comment by Tyriar** [comment]
vscode uses xterm, that's the issue that's been reported over there.

**Comment by gdlx** [comment]
Terminus has the same issue: https://github.com/Eugeny/terminus/issues/768

**Comment by ixrock** [comment]
It's still possible to catch `~` with `InputEvent`:

![image](https://user-images.githubusercontent.com/6377066/58793900-c2087680-85ff-11e9-8053-36d438e62c5b.png)

**Comment by Tyriar** [comment]
I'm not sure if this ever worked if `InputEvent` fixes the problem, sounds like it might be the same root cause as https://github.com/xtermjs/xterm.js/issues/469 and needs us to move input handling over to `input` instead of the key events.

**Comment by ixrock** [comment]
@Tyriar Using only `InputEvent` for sure not enough. Its `evt.data` just contains current value-string for `input/textarea` and nothing about pressed keys, etc. I just thought it might help to fix this issue combined with `KeyboardEvent`.

**Comment by DjouGoulart** [comment]
I'm having the same issue in [Fluent Terminal](https://github.com/felixse/FluentTerminal/). 
I can't type `~`, it shows `Þ` instead. My keyboard is PT-BR (with the AltGr key), but if I change the layout to US, I can type `~` using the `'` key.

**Comment by insertt** [comment]
Until this issue will be fixed I suggest using workaround from [here](https://superuser.com/questions/414405/pressing-the-tilde-key-waits-for-a-second-key-stroke-how-do-i-disable-this)

**Comment by GitSquared** [comment]
I'm not sure if this exactly the same issue, but I find that typing any diacritics (like ö ầ etc) on `xterm` doesn't work at all (prints like o a, without any diacritic applied). Tested with french keyboard and Linux, latest xTerm.js on Hyper and/or eDEX-UI.

**Comment by Tyriar** [comment]
@GitSquared I think this might be related to your `$LANG` variable?

**Comment by GitSquared** [comment]
@Tyriar My `$LANG` is set to `en_US.UTF-8` - but typing diacritics does work in other terminal emulators (Konsole), and switching to, for instance, `fr_FR.UTF-8` does not resolve the issue.

Also, I'm unsure how `$LANG` affects typing input? Is this some xterm.js-specific behavior?

**Comment by jerch** [comment]
@Tyriar Is this only an issue under Mac or with Mac keyboards (MBP)? I have no problems with PC keyboards with german keyboard layout - all third level shifts work as intented in all browsers (tested with demo in chrome + FF).

Some ideas whats going wrong (in fact it looks like 2 separate issues on the lower level):
- original poster reported to press `Alt + ^`: `Alt` has a special meaning by default, we map it to the `Meta` key for PC keyboards (as it lacks a dedicated `Meta` key). This assumption might be wrong for some keyboard types (Mac?). For Mac keyboards we have an `macOptionIsMeta` setting, we might have to check if this needs a slightly different `Alt+XY` key handling in `Keyboard.ts` as well if set.
- fluentTerminal and AltGr issue: This looks like the browser engine reports the wrong key/keyCode (not applying keyboard layout correctly?). When `AltGr` is held down normally the reported entry should shift to the third character for the second key, if browser do wrong here I fear there is not much we can do in the first place. But atm this is all speculation and needs proper bug testing to catch all edge cases. If `InputEvent` still reports the right key value we might get away with a combined `KeyboardEvent` + `InputEvent` processing. If this does not work - a last resort might be to deliver our own (customizable) keyboard layout maps.

Key handling always has been a pain in the a** in browsers, maybe we can learn something from the monaco guys here?

Things mentioned in this issue are hard to fix as long as we cannot repro it and dont know the exact circumstances this bug shows up (OS + browser version). Any volunteers to address (partly) this?

**Comment by laarmen** [comment]
FYI, a string of closed/duplicate issues on the VSCode repository lead me to this issue, and I'm on Windows. Using a French AZERTY layout, the tilde is on the "number" row, on the "2" key, which has as main char "é", shift char "2" and third char "~", which normally has a dead key behaviour. Thus, to type in a ~ character, the normal sequence would be AltGr+é, Space. However, when doing so the result is `é~`.

I'm not entirely certain it is the exact same bug, but I'm assuming it stems from the same root causes.

**Comment by laarmen** [comment]
OK, so I tried to reproduce on a Docker xterm.js instance, and couldn't reproduce (Firefox and Chrome both). I'm on Windows 10, and my bug happens in the VSCode terminal (1.38.1). 🤷‍♂

**Comment by jerch** [comment]
@laarmen Oh thats weird, Im pretty sure vscode runs exactly the same code for key input unaltered. 

@Tyriar FYI. Maybe an electron bug not correctly applying OS keyboard layout?

**Comment by Tyriar** [comment]
@rebornix have you seen any reports around this for monaco?

**Comment by ssh-peppe** [comment]
For me on macOS, Fnnish keyboard, with Safari and Chrome a tilde + space results in ~ and tilde + n results in ñ as it should. But on Firefox a tidle + space results in ~~and tidle + n results in ~ññ.

Interestingly this behavior difference can be seen (but a bit differently for some reason) on the https://xtermjs.org/ home page example terminal. If you enter tidle + space there on Safari you get nothing at all. Same for tidle + n. But with Firefox while tidle + space gets you nothing a tidle + n gets you ñ. Not sure why it's working differently there.

Debug logging shows the following when typing tidle + space on Firefox:
xterm.js: sending data "~" Array [ 126 ]
xterm.js: sending data "~" Array [ 126 ]
xterm.js: parsing data ~~

For tidle + n it shows:
xterm.js: sending data "~" Array [ 126 ]
xterm.js: sending data "ñ" Array [ 241 ]
xterm.js: sending data "ñ" Array [ 241 ]
xterm.js: parsing data ~
xterm.js: parsing data ññ

Note that there are differences in browsers with regards to what is reported from e.g. compositionend vs. input vs. keydown vs. keypress events.

**Comment by ssh-peppe** [comment]
The sequence of keyboard events recording by attaching a custom keyboard handler using the Terminal API, for tidle + n is on Windows Firefox (71) with Finnish keyboard is the following:

```
keydown Control { target: textarea.xterm-helper-textarea, key: "Control", charCode: 0, keyCode: 17 }
keydown { target: textarea.xterm-helper-textarea, key: "AltGraph", charCode: 0, keyCode: 18 }
keydown { target: textarea.xterm-helper-textarea, key: "Dead", charCode: 0, keyCode: 160 }
keyup { target: textarea.xterm-helper-textarea, key: "Dead", charCode: 0, keyCode: 160 }
keyup { target: textarea.xterm-helper-textarea, key: "Control", charCode: 0, keyCode: 17 }
keyup { target: textarea.xterm-helper-textarea, key: "AltGraph", charCode: 0, keyCode: 18 }
keydown { target: textarea.xterm-helper-textarea, key: "ñ", charCode: 0, keyCode: 78 }
keyup { target: textarea.xterm-helper-textarea, key: "n", charCode: 0, keyCode: 78 }
```

On macOS with Firefox (71) with Finnish keyboard the sequence is:
```
keydown Alt { target: textarea.xterm-helper-textarea, key: "Alt", charCode: 0, keyCode: 18 }
keydown Alt { target: textarea.xterm-helper-textarea, key: "Dead", charCode: 0, keyCode: 160 }
keyup Alt { target: textarea.xterm-helper-textarea, key: "~", charCode: 0, keyCode: 160 }
keyup { target: textarea.xterm-helper-textarea, key: "Alt", charCode: 0, keyCode: 18 }
keydown { target: textarea.xterm-helper-textarea, key: "ñ", charCode: 0, keyCode: 78 }
keyup { target: textarea.xterm-helper-textarea, key: "n", charCode: 0, keyCode: 78 }
```

**Comment by ssh-peppe** [comment]
Perhaps this is a different issue so I'll file a separate ticket.

**Comment by tcardonne** [comment]
Hello,

I'm having the same issue on VSCode's integrated terminal with a french layout.
When I want to input the `~` character, it prints me `é~` (`Alt right` + `é` then `space`).
`Alt right` + `é` (twice), it prints me `éé~`.
I can reproduce the issue on xtermjs.org.

I tried to gather some logs/infos to help you. If I understand correctly, you can't reproduce the issue on your side, so tell me if you need anything else.

VSCode debug, attached onkeydown and oninput to console.log :
![Capture2](https://user-images.githubusercontent.com/2379631/83577625-ff665380-a534-11ea-82df-71b24f1141f4.PNG)

(the é appears on AltGraph keyboard event).

Keyboard event viewer :
![Capture](https://user-images.githubusercontent.com/2379631/83577628-01301700-a535-11ea-9636-8ed4d928140e.PNG)

**Comment by Tyriar** [comment]
@tcardonne I believe this never worked and is happening because we're listening using keydown/key/keyup events as opposed to the `input` event, this is the same reason emoji support isn't great https://github.com/xtermjs/xterm.js/issues/469. This change might end up being relatively easy to move over but since it's a pretty fundamental change to how input works it will need a lot of validation and tests.

**Comment by tcardonne** [comment]
@Tyriar I understand.

It's weird, I have the issue on my desktop at home but I can't reproduce it with my work laptop (with the same keyboard attached to it).
I've tested on xtermjs.org website and on vscode.
The only difference appears to be the Windows version (2004 on personal computer, 1903 on work laptop).

Here is the keyboard event viewer table : 
![Capture](https://user-images.githubusercontent.com/2379631/83656454-dccd4c80-a5bf-11ea-9c88-0729117c36fd.PNG)
(notice the `Dead` key)

This might be related to a difference in OS settings, but I can't find any difference. Do you have any idea on settings that might affect this behavior?

**Comment by tcardonne** [comment]
Ok, I feel stupid, I didn't noticed the NumLock on my first comment.
I have a small keyboard without numpad and function keys, so I didn't noticed.

However, even with num lock, it should work, so the underlying issue (use of the input event) is still present, I guess.

Nevertheless, I'm happy to have found the cause of my issue.

**Comment by Zincr0** [comment]
> > I see that the [microsoft/vscode#98533](https://github.com/microsoft/vscode/issues/98533) was marked as duplicated for this, but https://xtermjs.org/ can type any áéíóú characters whitout any problems while the integrated vscode terminal cannot...
> 
> Disagree. I can't type any áéíóú characters in the https://xtermjs.org/ demo with my Spanish keyboard, neither.

Maybe is a SO or a browser thing? i'm using firefox on kde Neon here

**Comment by ricpelo** [comment]
> > > I see that the [microsoft/vscode#98533](https://github.com/microsoft/vscode/issues/98533) was marked as duplicated for this, but https://xtermjs.org/ can type any áéíóú characters whitout any problems while the integrated vscode terminal cannot...
> > 
> > 
> > Disagree. I can't type any áéíóú characters in the https://xtermjs.org/ demo with my Spanish keyboard, neither.
> 
> Maybe is a SO or a browser thing? i'm using firefox on kde Neon here

You're right: I can reproduce the issue using Google Chrome but the áéíóú characters works right on Firefox. :open_mouth:

**Comment by jerch** [comment]
@ricpelo This smells like an issue with "dead" keys behaving differently in FF, Chrome and Safari. Could you try to record the events with https://w3c.github.io/uievents/tools/key-event-viewer.html in FF and Chrome and post the results here?

**Comment by ricpelo** [comment]
@jerch sure. Here I go:

## Chrome:

![chrome](https://user-images.githubusercontent.com/616169/84427190-cd37ae80-ac24-11ea-9aa2-afffd117aa14.png)

## Firefox:

![firefox](https://user-images.githubusercontent.com/616169/84427198-ce68db80-ac24-11ea-871b-88f651867552.png)

**Comment by jerch** [comment]
@ricpelo Thx alot. The problem comes from this:

While FF already gives the dead key accumulated character on final `keydown` in `key`, chrome just returns the raw key. I am not sure who is to blame here, imho the keyboard stuff in `keydown` is not well specced at all. It also shows that `input` gets the chars right in both cases. Thats quite a strong argument to move to `input` (we already figured that above).

But `input` has difficulties to correctly catch meta keys like Strg, Alt, AltGr (left/right option on OSX), esp. Shift in combination, but we need those in a terminal to promote any Strg+X and such. I guess we need a more complex event handler, that basically relies on `input` but does another decoration/translation of the char from meta key states states of "enclosing" `keydown` / `keyup` events. Sounds nasty, but should be doable. Still the testing is hard to do (no good way to automate this without physical interaction).

**Comment by ricpelo** [comment]
Any news about this issue? Is there a proposed pull request or something similar?

**Comment by skneko** [comment]
> 
> 
> I see that the [microsoft/vscode#98533](https://github.com/microsoft/vscode/issues/98533) was marked as duplicated for this, but https://xtermjs.org/ can type any áéíóú characters whitout any problems while the integrated vscode terminal cannot...

Agree. That website works perfectly with my Spanish keyboard, but in VSCode I can't input tilde nor accented letters.

**Comment by ojacques** [comment]
🥳 The workaround mentioned earlier https://github.com/xtermjs/xterm.js/issues/2151#issuecomment-515116994 by @insertt worked for me (French keyboard). No more `é~` !

- Create / install the layout without dead keys per the instructions above (here is the one I built for me: [French Keyboard layout for VSCode.zip](https://github.com/xtermjs/xterm.js/files/6195688/French.Keyboard.layout.for.VSCode.zip))
- Reboot (otherwise, Win 10 did not took fully into account)
- In Win 10 / Language settings, add the new keyboard layout 

![image](https://user-images.githubusercontent.com/5385290/112284104-25f2a800-8c89-11eb-8db8-fcbcfa1de5f8.png)
(My PC is configured for US, but I use a French Keyboard)

**Comment by alexandrehpiva** [comment]
As @Tyriar set [this issue](https://github.com/microsoft/vscode/issues/105051) as a duplicate of this one, I am posting here that nothing solves for me.

I tried to re-install VSCode, change terminal to use ZSH shell, change the VSCode language and added configuration "terminal.integrated.env.linux" to use UTF-8.

I personally don't think it is the same problem of [this issue](https://github.com/microsoft/vscode/issues/105051) because I am using Ubuntu and when I type (in VSCode terminal) a special char alone and press space it works ex.: "^~`'", but when I type a special char like "^" + "a" only show "a" instead of "â". Same with "áâãéẽíĩóú". But in normal Ubuntu terminal everything works ok.

Does anyone have any more ideas?

**Comment by jerch** [comment]
@alexandrehpiva Can repro that under ubuntu with chrome and QWERTZ (german layout), where ^ is a dead key. I think dead key composition never worked correctly for chrome under linux, while it works with firefox.

Can you confirm - is ^ a dead key in your keyboard layout? (dead key - not producing a symbol on first press, instead adding some modification to a following key, often used to add ^, ~, ¨ or other accents to a letter)

**Comment by alexandrehpiva** [comment]
> Can you confirm - is ^ a dead key in your keyboard layout? (dead key - not producing a symbol on first press, instead adding some modification to a following key, often used to add ^, ~, ¨ or other accents to a letter)

No, it isn't. My keyboard is pt-BR Brazilian Portuguese. Really, my problem explanation is like the 
issue https://github.com/microsoft/vscode/issues/105051

**Comment by fperez** [comment]
@alexandrehpiva do you still get the problem on VSCode? I just tested it on my current setup (VSCode 1.59.1 on MacOS Big Sur, with US International layout) and it's now fine. Just checking to see if with current versions of VSCode it still happens, to help narrow down things.

**Comment by alexandrehpiva** [comment]
> @alexandrehpiva do you still get the problem on VSCode? I just tested it on my current setup (VSCode 1.59.1 on MacOS Big Sur, with US International layout) and it's now fine. Just checking to see if with current versions of VSCode it still happens, to help narrow down things.

Yes, I just update to 1.60.0, and still same problem. I'm unable to use accented characters with Ubuntu bash inside VSCode. I tried to install and use Oh My ZSH too and still same. The mad thing is that work's perfectly in terminal outside VSCode. Anyone knows how to change configuration for this (if exists)  in Ubuntu 20.04 LTS?

**Comment by alexandrehpiva** [comment]
I record an example:

![Peek 2021-09-03 08-40](https://user-images.githubusercontent.com/32683082/131999800-889864b0-9083-4f67-8565-3edfb9bdb20f.gif)

**Comment by alexandrehpiva** [comment]
I just discover a strange thing: If I press 3 times the accent key and then press the letter, works and i get: "~ã"


![Peek 2021-09-03 08-43](https://user-images.githubusercontent.com/32683082/132000295-98cf73dd-74a9-4545-bebe-3148d57b4c16.gif)

**Comment by thiagobraga** [comment]
> I just discovered a strange thing: If I press the accent key 3 times and then press the letter, it works and I get: "~ã"

I realized the same thing today. If I press the tilde key followed by the "a" key, it is displayed `a` in the terminal and not `ã`. But if I press the tilde three times and then the "a" key, `~ã` is produced. I imagine the issue should not be closed yet. I'm on Ubuntu 20.04.3 and on Terminator and Gnome Terminal the accent works perfectly.

**Comment by devphobia** [comment]
Ubuntu 20.04 using GNOME and having the same issues.
The integrated terminal doesn't seem to allow me to input characters like [é, ó, ú, ô, etc...] while all my other terminals work just fine. Even installed Atom to see if the bug persisted on other code editors but it seems it's related to VSCode and VSCode only. 
Any ideas on how to get rid of this?

**Comment by mauriciosantosds** [comment]
Same problem here with vscode integrated terminal in Ubuntu 20.04 GNOME.

**Comment by renathoaz** [comment]
For those using Ubuntu 20.04.
After a long searching I found out that vscode  depends on ibus for the accents work on integrated terminal.
the workaround :

1.Install ibus: ` sudo apt install ibus ibus-gtk3`
2. Create the config file:
```
cat <<EOF >ibus.dconf
[general]
preload-engines=['xkb:br::por']

[general/hotkey]
triggers=@as []

[panel]
show-icon-on-systray=false
EOF
```
3. import it to dconf:  `dconf load /desktop/ibus/<ibus.dconf`
4. init ibus with proper config: ` im-config -n ibus`
5. activate ibus on GNOME: `gsettings set org.gnome.desktop.input-sources sources "[('ibus', 'xkb:br::por')]"`

### and now everything is just working fine:
![Seleção_329](https://user-images.githubusercontent.com/16307350/160263549-6f699dab-f9af-462e-8884-fe8f07d8cbac.png)

**Comment by thiagobraga** [comment]
Now I'm not using Ubuntu 20.04 anymore, I upgraded to Ubuntu 21.10 and VSCode 1.66. I noticed that `ibus` and `ibus-gtk3` are already installed. I believe that since I reinstalled, I hadn't done the accent test in VSCode terminal. And to my surprise, it now worked normally, without having to change any settings.

![image](https://user-images.githubusercontent.com/993684/161872122-aba33059-6ff5-4d18-8870-7769068dc6f4.png)
